### PR TITLE
Fix #13018: EqualsFilterConstraint handle enum/string

### LINKED ISSUE #13018: Datatable: Default filter feature broken in showcase with PF15
### Describe the bug

While doing some manual testing for the JS refactor PR, I noticed the default filter feature from the showcase is broken and throws an exception.

### Reproducer

* Fetch the lastest state of the master branch (I used ccfd6d682d8041fd01a7af50c0fdaadde91e8d10), start the showcase and goto the data table filter feature page: http://localhost:8081/showcase/ui/data/datatable/filter.xhtml
* Scroll down to `Default Filter`
* Attempt to use any of the filter. E.g. I clicked in the filter at the top of the `Country` column and entered `Spain`

This results in a generic error page with the following stacktrace:

```
Dec 08, 2024 2:03:26 PM org.primefaces.application.exceptionhandler.PrimeExceptionHandler logException
SEVERE: Filter 'NEW' of type 'java.lang.String' cannot be casted to value type 'NEGOTIATION'. Forgot to add a converter?
java.lang.IllegalArgumentException: Filter 'NEW' of type 'java.lang.String' cannot be casted to value type 'NEGOTIATION'. Forgot to add a converter?
	at org.primefaces.model.filter.ComparableFilterConstraint.assertAssignable(ComparableFilterConstraint.java:57)
	at org.primefaces.model.filter.EqualsFilterConstraint.isMatching(EqualsFilterConstraint.java:44)
	at org.primefaces.component.datatable.feature.FilterFeature.lambda$filter$0(FilterFeature.java:153)
	at org.primefaces.component.api.ColumnAware.forEachColumn(ColumnAware.java:102)
	at org.primefaces.component.api.ColumnAware.forEachColumn(ColumnAware.java:55)
	at org.primefaces.component.api.ColumnAware.forEachColumn(ColumnAware.java:51)
	at org.primefaces.component.datatable.feature.FilterFeature.filter(FilterFeature.java:132)
	at org.primefaces.component.datatable.feature.FilterFeature.encode(FilterFeature.java:94)
	at org.primefaces.component.datatable.DataTableRenderer.encodeEnd(DataTableRenderer.java:93)
	at jakarta.faces.component.UIComponentBase.encodeEnd(UIComponentBase.java:586)
	at jakarta.faces.component.UIComponent.encodeAll(UIComponent.java:1442)
	at com.sun.faces.context.PartialViewContextImpl$PhaseAwareVisitCallback.visit(PartialViewContextImpl.java:624)
	at com.sun.faces.component.visit.PartialVisitContext.invokeVisitCallback(PartialVisitContext.java:158)
	at jakarta.faces.component.visit.VisitContextWrapper.invokeVisitCallback(VisitContextWrapper.java:92)
	at org.primefaces.component.api.PrimeUIData$VisitContextImpl.invokeVisitCallback(PrimeUIData.java:508)
	at org.primefaces.component.patch.UIDataPatch.visitTree(UIDataPatch.java:877)
	at org.primefaces.component.api.PrimeUIData.visitTree(PrimeUIData.java:292)
	at jakarta.faces.component.UIComponent.visitTree(UIComponent.java:1278)
	at jakarta.faces.component.UIForm.visitTree(UIForm.java:344)
	at jakarta.faces.component.UIComponent.visitTree(UIComponent.java:1278)
	at jakarta.faces.component.UIComponent.visitTree(UIComponent.java:1278)
	at com.sun.faces.context.PartialViewContextImpl.processComponents(PartialViewContextImpl.java:394)
	at com.sun.faces.context.PartialViewContextImpl.processPartial(PartialViewContextImpl.java:312)
	at org.primefaces.context.PrimePartialViewContext.processPartial(PrimePartialViewContext.java:61)
	at jakarta.faces.component.UIViewRoot.encodeChildren(UIViewRoot.java:1092)
	at jakarta.faces.component.UIComponent.encodeAll(UIComponent.java:1435)
	at com.sun.faces.application.view.FaceletViewHandlingStrategy.renderView(FaceletViewHandlingStrategy.java:422)
	at com.sun.faces.application.view.MultiViewHandler.renderView(MultiViewHandler.java:160)
	at jakarta.faces.application.ViewHandlerWrapper.renderView(ViewHandlerWrapper.java:125)
	at jakarta.faces.application.ViewHandlerWrapper.renderView(ViewHandlerWrapper.java:125)
	at com.sun.faces.lifecycle.RenderResponsePhase.execute(RenderResponsePhase.java:93)
	at com.sun.faces.lifecycle.Phase.doPhase(Phase.java:72)
	at com.sun.faces.lifecycle.LifecycleImpl.render(LifecycleImpl.java:150)
	at jakarta.faces.lifecycle.LifecycleWrapper.render(LifecycleWrapper.java:91)
	at jakarta.faces.webapp.FacesServlet.executeLifecyle(FacesServlet.java:692)
	at jakarta.faces.webapp.FacesServlet.service(FacesServlet.java:449)
	at org.eclipse.jetty.ee10.servlet.ServletHolder$NotAsync.service(ServletHolder.java:1379)
	at org.eclipse.jetty.ee10.servlet.ServletHolder.handle(ServletHolder.java:736)
	at org.eclipse.jetty.ee10.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1614)
	at org.primefaces.webapp.filter.NoCacheFilter.doFilter(NoCacheFilter.java:50)
	at org.eclipse.jetty.ee10.servlet.FilterHolder.doFilter(FilterHolder.java:205)
	at org.eclipse.jetty.ee10.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1586)
	at org.primefaces.webapp.filter.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:53)
	at org.eclipse.jetty.ee10.servlet.FilterHolder.doFilter(FilterHolder.java:208)
	at org.eclipse.jetty.ee10.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1586)
	at org.eclipse.jetty.ee10.websocket.servlet.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:195)
	at org.eclipse.jetty.ee10.servlet.FilterHolder.doFilter(FilterHolder.java:205)
	at org.eclipse.jetty.ee10.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1586)
	at org.eclipse.jetty.ee10.servlet.ServletHandler$MappedServlet.handle(ServletHandler.java:1547)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.dispatch(ServletChannel.java:819)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.handle(ServletChannel.java:436)
	at org.eclipse.jetty.ee10.servlet.ServletHandler.handle(ServletHandler.java:464)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:575)
	at org.eclipse.jetty.ee10.servlet.SessionHandler.handle(SessionHandler.java:717)
	at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1060)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:151)
	at org.eclipse.jetty.server.Server.handle(Server.java:182)
	at org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:662)
	at org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:418)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:478)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:441)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:293)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run(AdaptiveExecutionStrategy.java:201)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:311)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:979)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1209)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1164)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

### Expected behavior

No error page should show up, the filter should be applied to the database.  Not sure if PF or the showcase needs to be updated. It works in the [current PF 14 showcase](https://www.primefaces.org/showcase/ui/data/datatable/filter.xhtml)

### PrimeFaces edition

None

### PrimeFaces version

ccfd6d682d8041fd01a7af50c0fdaadde91e8d10

### Theme

_No response_

### JSF implementation

All

### JSF version

2.3.21

### Java version

17

### Browser(s)

_No response_

### PR DESCRIPTION
**Pull request by melloware**
Fix #13018: EqualsFilterConstraint handle enum/string

### CODE REVIEW

File: primefaces/src/main/java/org/primefaces/model/filter/EqualsFilterConstraint.java [hunk 1]
```diff
@@ -39,6 +39,10 @@ public boolean isMatching(FacesContext ctxt, Object value, Object filter, Locale
         if (value == null || filter == null) {
             return false;
         }
+        // Handle enum values by comparing their string representations
+        if (value.getClass().isEnum()) {
+            return value.toString().equals(filter.toString());
```
Comment (tandraschko) [inline primefaces/src/main/java/org/primefaces/model/filter/EqualsFilterConstraint.java:44]: doesnt `value.equals(filter)` work?

**Comment by melloware** [comment]
I will test today and I think I have some more unit tests to add

**Comment by melloware** [comment]
Done. The NULL checks are actually done above it and it breaks out if either is NULL.

```java
if (value == null || filter == null) {
            return false;
        }
```

**Comment by tandraschko** [comment]
+1 then

**Pull request merged**